File: DynamicPrograming/csharp/SumPossible.cs
using System.Collections.Generic;

namespace DynamicProgrammingSolutions;

public static class SumPossible
{
    public static bool IsPossible(int amount, IReadOnlyList<int> numbers)
    {
        var memo = new Dictionary<int, bool>();
        return Dfs(amount, numbers, memo);
    }

    private static bool Dfs(int remaining, IReadOnlyList<int> numbers, IDictionary<int, bool> memo)
    {
        if (remaining == 0)
        {
            return true;
        }

        if (remaining < 0)
        {
            return false;
        }

        if (memo.TryGetValue(remaining, out var cached))
        {
            return cached;
        }

        foreach (var num in numbers)
        {
            if (Dfs(remaining - num, numbers, memo))
            {
                memo[remaining] = true;
                return true;
            }
        }

        memo[remaining] = false;
        return false;
    }
}

File: DynamicPrograming/java/SumPossible.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class SumPossible {
  private SumPossible() {
  }

  public static boolean isPossible(int amount, List<Integer> numbers) {
    Map<Integer, Boolean> memo = new HashMap<>();
    return dfs(amount, numbers, memo);
  }

  private static boolean dfs(int remaining, List<Integer> numbers, Map<Integer, Boolean> memo) {
    if (remaining == 0) {
      return true;
    }
    if (remaining < 0) {
      return false;
    }
    if (memo.containsKey(remaining)) {
      return memo.get(remaining);
    }
    for (int num : numbers) {
      if (dfs(remaining - num, numbers, memo)) {
        memo.put(remaining, true);
        return true;
      }
    }
    memo.put(remaining, false);
    return false;
  }
}
